### ISSUE TITLE
CMake FMU export produces too long paths

### ISSUE THREAD

**Issue by bilderbuchi** [OpenModelica/OpenModelica#12916]
When exporting FMUs of models with long names, e.g. the MSL `Modelica.Magnetic.QuasiStatic.FluxTubes.Examples.BasicExamples.ToroidalCoreQuadraticCrossSection`, using today's nightly, one now gets more output from the build process. In this case, CMake warns me in the "generate target files" step:
```
CMake Warning in CMakeLists.txt:
  The object file directory

    C:/Users/my_user/AppData/Local/Temp/OpenModelica/OMEdit/ToroidalCoreQuadraticCrossSection5ba5/ToroidalCoreQuadraticCrossSection.fmutmp/sources/build_cmake/CMakeFiles/ToroidalCoreQuadraticCrossSection.dir/

  has 205 characters.  The maximum full path to an object file is 250
  characters (see CMAKE_OBJECT_PATH_MAX).  Object file

    simulation/solver/initialization/initialization.c.obj

  cannot be safely placed under this directory.  The build may not work
  correctly.
```
I'm not sure if that messes with the build? 
There is some duplication here, but it's only the model name, so I'm not sure how to best mitigate this further?
* Drop `OpenModelica` (which is an empty folder) from the working directory path? This might not be that attractive, I don't know if other OM tools would put stuff in there?
* use just the hash for the last folder level in `CMakeFiles` -- `CMakeFiles/12d2/`? That folder seems to be not _that_ user-visible/-important (as opposed to the `.fmutmp` one)

OM OpenModelica v1.25.0-dev-31, Windows 10.
Originally from https://github.com/OpenModelica/OpenModelica/issues/12828#issuecomment-2370403969

**Comment by arun3688** [comment]
@bilderbuchi This is a well known issue with long paths, and it affects the builds. We tried to reduce the path by just using the modelNames along with hash may be that is not sufficient and we need to reduce it further it seems, Can you just try to change the working directory to some short path and test it and it should work fine. Apart from this long path issue, i hope fmu export should be working fine, i will discuss with @adeas31 on reducing the path length and get back to you. But until then you can try to switch to some short working directory or try to extend the model with short names.

**Comment by bilderbuchi** [comment]
> But until then you can try to switch to some short working directory or try to extend the model with short names.

Yes, that should be a feasible workaround, thanks!

**Comment by arun3688** [comment]
> Yes, that should be a feasible workaround, thanks! I specifically searched for an MSL model that shows the issue, so that it's not just "my fault" for using too long model names. 😁

Absolutely We can avoid this long paths for example `C:/Users/u1ac_cbu/AppData/Local/Temp/OpenModelica/OMEdit/ToroidalCoreQuadraticCrossSection5ba5/ToroidalCoreQuadraticCrossSection.fmutmp/sources/build_cmake/CMakeFiles/ToroidalCoreQuadraticCrossSection.dir`

eliminating the `modelName` and just use only the `hash` to create the directory which would reduce around 40 to 60 chars for this example and the path would be just something like this `C:/Users/u1ac_cbu/AppData/Local/Temp/OpenModelica/OMEdit/5ba5/5ba5.fmutmp/sources/build_cmake/CMakeFiles/ToroidalCoreQuadraticCrossSection.dir`

**Comment by bilderbuchi** [comment]
Yeah, that's similar to what I proposed above. Just keep user experience in mind! 
For examples, in this case I would retain the model name instead of `.../OMEdit/5ba5/`, because otherwise the working directory will become filled with hash-only dirnames and it will be hard to find the model you are working on.
I'd say that the one in `CMakeFiles/ToroidalCoreQuadraticCrossSection.dir` could be hashed, because at that point it's clear from the rest of the path in which model you are, and I don't know how often users browse there. Similar, but a _little_ less clear-cut for the fmutmp.

So, from a discoverability point-of-view, and with maximum shortening, I'd go with `C:/Users/my:user/AppData/Local/Temp/OpenModelica/OMEdit/ToroidalCoreQuadraticCrossSection5ba5/5ba5.fmutmp/sources/build_cmake/CMakeFiles/5ba5.dir`. 
The "top" level ties the name and the hash together for clarity, and the .fmutmp and .dir paths just use the hash for shortness. 
What do you think?

**Comment by casella** [comment]
@bilderbuchi  @arun3688 we already tackled something similar in #11554. Wasn't that enough? Do I miss something?

**Comment by arun3688** [comment]
> @bilderbuchi @arun3688 we already tackled something similar in #11554. Wasn't that enough? Do I miss something?

This is true, but still OMEdit creates additional directory with `modelName + hash` and also it creates `OpenModelica/OMEdit` in the temp directory i had discussion with adeel regarding to eliminate few stuffs, but we thought that the creating directory with modelname + hash, instead of full MSL name solved the issue, but unfortunately we still have models with very large names

**Comment by casella** [comment]
> So, from a discoverability point-of-view, and with maximum shortening, I'd go with `C:/Users/my_user/AppData/Local/Temp/OpenModelica/OMEdit/ToroidalCoreQuadraticCrossSection5ba5/5ba5.fmutmp/sources/build_cmake/CMakeFiles/5ba5.dir`. The "top" level ties the model name and the hash together for clarity (as it is now), and the .fmutmp and .dir paths just use the hash for shortness. What do you think?

LGTM. Non need to replicate the model name everywhere, particularly in temporary directories that are only used to store intermediate files during the FMU build process.

**Comment by arun3688** [comment]
@bilderbuchi the issue is fixed with this 29c19a45bc9c574bc8029844166a2bcf6f07b7aa, So now we use hash names with length of 3 (e.g) `ToroidalCoreQuadraticCrossSection5ba5.fmutmp` to something like `257.fmutmp`, i tested the above failing example from OMEdit with working directory set to `C:/Users/my_user/AppData/Local/Temp/OpenModelica/OMEdit/` and it works, please test it from the nightly builds tomorrow (1-11-2024)
